src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h

```diff
@@ -0,0 +1,1729 @@
+  }
+
+  /**
+   * copy_from_foreign
+   *
+   * Copies entries from [from_src, to_src) to tgt.
+   *
+   * tgt and from_src must be from different nodes.
+   * from_src and to_src must be from the same node.
+   */
+  static void copy_from_foreign_head(
+    iterator tgt,
+    const_iterator from_src,
+    const_iterator to_src) {
+    assert(tgt->node != from_src->node);
+    assert(to_src->node == from_src->node);
+    void* des = tgt.node->from_end((to_src -1)->get_node_key().key_off);
+    void* src = (to_src - 1)->get_node_val_ptr();
+    size_t len = (to_src -1)->get_node_key().key_off;
+    memcpy(des, src, len);
+    memcpy(
+      tgt->get_node_key_ptr(), from_src->get_node_key_ptr(),
+      to_src->get_node_key_ptr() - from_src->get_node_key_ptr());
+  }
+
+  void copy_from_foreign_back(
+    iterator tgt,
+    const_iterator from_src,
+    const_iterator to_src) {
+    assert(tgt->node != from_src->node);
+    assert(to_src->node == from_src->node);
+    auto offset = from_src.get_index() == 0? 0: (from_src-1)->get_node_key().key_off;
+    void* des = tgt.node->from_end((to_src -1)->get_node_key().key_off - offset);
+    void* src = (to_src - 1)->get_node_val_ptr();
+    size_t len = from_src.get_index() == 0? (to_src -1)->get_node_key().key_off:
+                 (from_src-1)->get_node_val_ptr() - (to_src -1)->get_node_val_ptr();
+    memcpy(des, src, len);
+    memcpy(
+      tgt->get_node_key_ptr(), from_src->get_node_key_ptr(),
+      to_src->get_node_key_ptr() - from_src->get_node_key_ptr());
+    if ( from_src.get_index() == 0)
+      return;
+
+    omap_inner_key_t key = (from_src - 1)->get_node_key();
+    auto end_idx = tgt.get_index() + to_src.get_index() - from_src.get_index();
+    for (auto ite = tgt; ite.get_index() != end_idx; ite++) {
+       omap_inner_key_t node_key = ite->get_node_key();
+       node_key.key_off -= key.key_off;
+       ite->set_node_key(node_key);
+    }
+  }
+
+  void append_copy_from_foreign_head(
+    iterator tgt,
+    const_iterator from_src,
+    const_iterator to_src) {
+    assert(tgt->node != from_src->node);
+    assert(to_src->node == from_src->node);
+    if (from_src == to_src)
+      return;
+
+    void* des = tgt.node->from_end((to_src -1)->get_node_key().key_off + (tgt - 1)->get_node_key().key_off);
+    void* src = (to_src - 1)->get_node_val_ptr();
+    size_t len = (to_src -1)->get_node_key().key_off;
+    memcpy(des, src, len);
+    memcpy(
+      tgt->get_node_key_ptr(), from_src->get_node_key_ptr(),
+      to_src->get_node_key_ptr() - from_src->get_node_key_ptr());
+    omap_inner_key_t key = (tgt - 1)->get_node_key();
+    auto end_idx = tgt.get_index() + to_src.get_index() - from_src.get_index();
+    for (auto ite = tgt; ite.get_index() != end_idx; ite++) {
+       omap_inner_key_t node_key = ite->get_node_key();
+       node_key.key_off += key.key_off;
+       ite->set_node_key(node_key);
+    }
+  }
+
+  /**
+   * copy_from_local
```

Comment:
Comment is wrong.  Whats the difference between local_move_back and local_move_ahead?  Please check the other comments in this file as well.